### PR TITLE
LogField

### PR DESCRIPTION
**Pull request by LeanderK**
The pull-request handles issue #54.

### CODE REVIEW

File: numhask/src/NumHask/Data/LogField.hs [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+-- Because 'logField' performs the semantic conversion, we can use
+-- operators which say what we *mean* rather than saying what we're
+-- actually doing to the underlying representation. That is,
+-- equivalences like the following are true[1] thanks to type-class
+-- overloading:
+--
+-- > logField (p + q) == logField p + logField q
+-- > logField (p * q) == logField p * logField q
+--
+-- TODO: what to do with this note?
```
Comment (tonyday567) [inline numhask/src/NumHask/Data/LogField.hs]: I think we briefly summarise the motivation of the original library, credit it, and just say that this is the numhask version.

File: numhask/src/NumHask/Data/LogField.hs [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+-- boundary twice; with a @LogField@ and a 'Double' it's three
+-- times, since the regular number needs to enter the log-domain
+-- first. This makes addition incredibly slow. Again, if you can
+-- parenthesize to do normal-domain operations first, do it!
+--
+-- [1] That is, true up-to underflow and floating point fuzziness.
+-- Which is, of course, the whole point of this module.
+newtype LogField a = LogField a
+      deriving (Eq, Ord, Read, Data, Generic, Generic1, Functor, Foldable, Traversable)
+    -- TODO original has comment: , Ord -- Should we really perpetuate the Ord lie?
```
Comment (tonyday567) [inline numhask/src/NumHask/Data/LogField.hs]: No.  Also what about a Show instance?

File: numhask/src/NumHask/Data/LogField.hs [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+
+----------------------------------------------------------------
+-- | Constructor which does semantic conversion from normal-domain
+-- to log-domain. Throws errors on negative and NaN inputs. If @p@
+-- is non-negative, then following equivalence holds:
+--
+-- > logField p == logToLogField (log p)
+logField :: (ExpField a) => a -> LogField a
+{-# INLINE [0] logField #-}
+-- TODO: should we use NOINLINE or [~0] to avoid the possibility of code bloat?
```
Comment (tonyday567) [inline numhask/src/NumHask/Data/LogField.hs]: Maybe.  There are no inline pragmas in numhask yet, awaiting experimentation.

File: numhask/src/NumHask/Data/LogField.hs [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+instance (ExpField a, Show a) => Show (LogField a) where
+    showsPrec p (LogField x) =
+        let y = exp x in y `seq`
+        showParen (p > 9)
+            ( showString "LogField "
+            . showsPrec 11 y
+            )
+
+
+-- TODO: never worked with FFI, so i don't really know what's going on
```
Comment (tonyday567) [inline numhask/src/NumHask/Data/LogField.hs]: No idea either.  Let's ditch it all for now.  If I had to guess, there may be some tricky specialisation stuff we can do when working with doubles.

File: numhask/src/NumHask/Data/LogField.hs [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+"log1p/expm1"    forall x. log1p (expm1 x) = x
+    #-}
+-- #endif
+
+instance (ExpField a, LowerBoundedField a, Ord a) => AdditiveMagma (LogField a) where
+      (LogField x) `plus` (LogField y)
+            -- TODO unsure what to do here...currentl only negInfinity because of the 0-element
+            | x == y
+            && x == negInfinity
+            && y == negInfinity = LogField x -- @0+0 == 0@,
```
Comment (tonyday567) [inline numhask/src/NumHask/Data/LogField.hs]: = zero is fine for the first guard.  You could also test if either is zero.

```
x@(LogField x') `plus` y@(LogField y')
  | x == zero && y == zero = zero
  | x == zero = y
  | y == zero = x etc
```

File: numhask/src/NumHask/Data/LogField.hs [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+    , AdditiveInvertible a
+    , AdditiveRightCancellative a
+    , LowerBoundedField a
+    , Eq a) => MultiplicativeRightCancellative (LogField a)
+
+instance (Multiplicative (LogField a), AdditiveInvertible a, AdditiveGroup a, LowerBoundedField a, Eq a) => MultiplicativeGroup (LogField a)
+
+instance (LowerBoundedField a, ExpField a, Ord a, AdditiveMagma a) => Distribution (LogField a)
+
+-- TODO: okay?
```
Comment (tonyday567) [inline numhask/src/NumHask/Data/LogField.hs]: I'm not sure - I think so, but the testing phase tends to sort this stuff out.

File: numhask/src/NumHask/Data/LogField.hs (hunk 4 above)
Comment (LeanderK) [reply numhask/src/NumHask/Data/LogField.hs]: Probably, but I don't think this is the concern of the module

File: numhask/src/NumHask/Data/LogField.hs (hunk 5 above)
Comment (LeanderK) [reply numhask/src/NumHask/Data/LogField.hs]: ah! yeah, that's smart 😉

File: numhask/src/NumHask/Algebra/Metric.hs [hunk 7]
```diff
@@ -306,12 +307,13 @@ instance Metric Word64 Word64 where
...
 
 -- | todo: This should probably be split off into some sort of alternative Equality logic, but to what end?
-class (Eq a, AdditiveGroup a) =>
+class (Eq a, AdditiveUnital a) =>
       Epsilon a where
   nearZero :: a -> Bool
   nearZero a = a == zero
 
   aboutEqual :: a -> a -> Bool
+  default aboutEqual :: AdditiveGroup a => a -> a -> Bool
```
Comment (LeanderK) [inline numhask/src/NumHask/Algebra/Metric.hs]: @tonyday567 that's what I meant in the ticket. With this formulation, I am able to write an `Epsilon` instance for `Logfield`

**Comment by LeanderK** [comment]
@tonyday567 I have written tests. Theoretically ready to merge, but I don't know whether this is true in practice 😉 is there anything you want me to change?

**Comment by LeanderK** [comment]
ok, something's wrong. I'll probably find time in the next 5 hours to look into it.

**Comment by tonyday567** [comment]
Looking good to me!  I tend to hit the button if it's all green, and we can always clean up later.

Appreciate the PR.

**Comment by ocramz** [comment]
I've just seen `log-domain` https://hackage.haskell.org/package/log-domain , which however has a type that implements Num as well.
